library/payments.ts

```diff
@@ -0,0 +1,283 @@
+  // Update transaction table and user balance in one tx
+  return trx("transaction")
+    .update({
+      fee_msat: paymentData.data.fee,
+      payment_request: paymentData.data.invoice,
+      success: true,
+      is_pending: false,
+      external_id: paymentData.data.id,
+      preimage: paymentData.data.preimage,
+    })
+    .where({ id: paymentRecordId })
+    .then(() => {
+      // Decrement balance and unlock user
+      return trx("user")
+        .decrement({
+          msat_balance: msatAmount + parseInt(paymentData.data.fee),
+        })
+        .update({ updated_at: db.knex.fn.now(), is_locked: false })
+        .where({ id: userId });
+    })
+    .then(trx.commit)
+    .then(() => {
+      log.debug(`Payment successful for ${userId}`);
+      return res
+        ? res
+            .status(200)
+            .send({ invoiceId: paymentRecordId, ...paymentData.data })
+        : {
+            success: true,
+            data: {
+              invoiceId: paymentRecordId,
+              preimage: paymentData.data.preimage,
+              feeMsat: paymentData.data.fee,
+            },
+          };
+    })
+    .catch((err) => {
+      log.error(
+        `Error updating transaction table on handleCompletedPayment: ${err}`
+      );
+      return res
+        ? res.status(500).send("Payment succeeded but update failed")
+        : { success: false, error: "UNKNOWN" };
+    });
+}
+
+async function handleFailedPayment(
+  res: any, // TODO: Use express response type
+  userId: string,
+  paymentRecordId: number,
+  paymentData: sendPaymentResponse
+) {
+  const trx = await db.knex.transaction();
+  // Update transaction table and user status
+  return trx("transaction")
+    .update({
+      success: false,
+      is_pending: false,
+      failure_reason: paymentData.message,
+    })
+    .where({ id: paymentRecordId })
+    .then(() => {
+      // Unlock user
+      return trx("user")
+        .update({ updated_at: db.knex.fn.now(), is_locked: false })
+        .where({ id: userId });
+    })
+    .then(trx.commit)
+    .then(() => {
+      log.debug(`Payment failed for ${userId}`);
+      return res
+        ? res.status(400).send(`Payment failed: ${paymentData.message}`)
+        : { success: false, error: paymentData.message };
+    })
+    .catch((err) => {
+      log.error(
+        `Error updating transaction table on handleFailedPayment: ${err}`
+      );
+      return res
+        ? res.status(500).send("Update failed on failed payment")
+        : { success: false, error: "UNKNOWN" };
+    });
+}
+
+async function isDuplicatePayRequest(invoice) {
+  return new Promise((resolve, reject) => {
+    return db
+      .knex("transaction")
+      .select("payment_request")
+      .where("payment_request", "=", invoice)
+      .then((data) => {
+        resolve(data.length > 0);
+      })
+      .catch((e) => {
+        log.error(`Error looking up invoice: ${e}`);
+        reject();
+      });
```

Comment:
i think we can ditch the promise wrapper here too
```suggestion
  return db
      .knex("transaction")
      .select("payment_request")
      .where("payment_request", "=", invoice)
      .then((data) => {
        return data.length > 0;
      })
      .catch((e) => {
        log.error(`Error looking up invoice: ${e}`);
        return;
      });
```